Add Support For Passing Nonce and Configurable JWT Expiry

rpcope1 opened the pull request:
## Description
* Added support to the Python client to allow the nonce to be set in the auth URL
* Added a client parameter to allow JWT expiry to be configured.
* Updated some tests to check nonce validation.

## Motivation and Context
I am working on integrating the Duo Universal prompt into an existing Python application that already does OIDC auth flows, and utilizing the OIDC nonce and potentially setting a shorter expiry window is being considered. These are supported in the Duo OIDC doc, but not exposed in the Python client library, so this would add that.

## How Has This Been Tested?
* I've added a couple of unit tests.

## Types of Changes
- [?] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

rpcope1 committed:
Add Support For Passing Nonce and Configurable JWT Expiry

* Added support to the Python client to allow the nonce to be set in
  the auth URL
* Added a client parameter to allow JWT expiry to be configured.
* Updated some tests to check nonce validation.

AaronAtDuo commented on duo_universal/client.py:
```diff
@@ -106,14 +112,15 @@ def _create_jwt_args(self, endpoint):
             'sub': self._client_id,
             'aud': endpoint,
-            'exp': time.time() + FIVE_MINUTES_IN_SECONDS,
+            'exp': time.time() + self._exp_seconds,
```
I think we want to limit the maximum value here to the default 5 minutes.  Shortening the expiration is fine but I don't want to give applications the option to lengthen the expiration, since this is a primary mitigation against replay attacks.

AaronAtDuo commented on duo_universal/client.py:
```diff
@@ -126,6 +133,7 @@ def __init__(self, client_id, client_secret, host,
         duo_certs                -- (Optional) Provide custom CA certs
         use_duo_code_attribute   -- (Optional: default true) Flag to use `duo_code` instead of `code` for returned authorization parameter
         http_proxy               -- (Optional) HTTP proxy to tunnel requests through
+        exp_seconds              -- (Optional) The number of seconds used for JWT expiry
```
we could also clamp the expiration seconds here instead

AaronAtDuo commented on duo_universal/client.py:
```diff
@@ -198,21 +207,23 @@ def health_check(self):
         username        -- username trying to authenticate with Duo
         state           -- Randomly generated character string of at least 22
                            chars returned to the integration by Duo after 2FA
+        nonce           -- Randomly generated character string of at least 16
```
You say 16 characters here, but it looks like you're reusing the state parameter's limits (back at line 103), so is 16 correct, or the 22 that line 216 has?

rpcope1 replied:
I believe this SDK is backed by the Duo OIDC API, and in the docs (https://duo.com/docs/oauthapi#authorization-request) both the nonce:

" Optional 	If provided, this must be a random value that is cryptographically secure such that it is unguessable by a third party. We require a length of at least 16 characters and at most 1024 characters. "

and state: 

"A random, cryptographically secure value that the client must maintain throughout the authentication. We require a length of at least 16 characters and at most 1024 characters. "

appear to specify a length of 16 to 1024 characters. Perhaps the thing to do is to also have the documentation for state also reflect these limitations?

AaronAtDuo replied:
I agree.... this is the first SDK we wrote so maybe things hadn't been finalized yet.  I'll open an issue to check and correct the various limits.

AaronAtDuo replied:
Issue 23 opened

AaronAtDuo commented on duo_universal/client.py:
```diff
@@ -222,7 +233,7 @@ def create_auth_url(self, username, state):
             'iss': self._client_id,
             'aud': API_HOST_URI_FORMAT.format(self._api_host),
-            'exp': time.time() + FIVE_MINUTES_IN_SECONDS,
+            'exp': time.time() + self._exp_seconds,
```
same concern here re: clamping the expiration to a maximum 5 minutes

AaronAtDuo commented on tests/test_validate_create_auth_url_inputs.py:
```diff
@@ -44,6 +44,22 @@ def test_long_state(self):
+        Test _validate_create_auth_url_inputs
+        throws a DuoException if the nonce is set and is too short
+        """
+        with self.assertRaises(client.DuoException) as e:
```
Linter doesn't like the unused 'e' :(

AaronAtDuo replied:
CI still failing due to this :(

rpcope1 commented:
The suggestions make sense, I will integrate changes to reflect you comments above.

rpcope1 committed:
Configurable Nonce/Expiry -- PR Review Suggestions

* Client now throws an exception if an out of range expiry
  was given on initialization; the value is furthermore
  clamped to between 0 seconds and 5 minutes where it's used.
* Doc strings have been updated to reflect what the underlying
  OIDC API specifies around the state and nonce.
* Tests have been added to verify that the expiry is checked
  correctly.

rpcope1 commented:
I have just pushed another commit that should hopefully address the points you've mentioned in the code review.

AaronAtDuo commented:
@rpcope1 Thanks! This looks great.  Just need to fix the linter issue so the tests will run, then I think we can get this merged.

rpcope1 committed:
* Fixed flake8 errors on a test.

rpcope1 commented:
> @rpcope1 Thanks! This looks great. Just need to fix the linter issue so the tests will run, then I think we can get this merged.

I just pushed a change set that I think will address the linter errors. Thank you so much for your help with this.

AaronAtDuo commented:
I think enough has been added since the last release that it's worth putting one out.  I'll try to get that done early next week.

AaronAtDuo commented:
Thanks @rpcope1!  The PR is merged and I'll try to get a release out next week.
